pandas/io/parsers.py

```diff
@@ -1681,10 +1690,25 @@ def _infer_columns(self):
 
             columns = []
             for level, hr in enumerate(header):
-                line = self._buffered_line()
+                try:
+                    line = self._buffered_line()
 
-                while self.line_pos <= hr:
-                    line = self._next_line()
+                    while self.line_pos <= hr:
+                        line = self._next_line()
+
+                except StopIteration:
+                    if self.line_pos < hr:
+                        raise StopIteration(
```

Comment:
I would just change these to ``ValueError`` (and fix the issue in the html/excel parser, should just mean catching a different exception). This would greatly align the c/python parser errors. (same on line 1709)